### PR TITLE
redirect to homepage if not authenticated, added redirect route for /…

### PR DESCRIPTION
**Pull request by RnbWd**
…campaigns/:id/milestones to prevent dead url, modify back button behavior

fixes https://github.com/Giveth/giveth-dapp/issues/329, https://github.com/Giveth/giveth-dapp/issues/182,

### CODE REVIEW

**Comment by vojtechsimetka** [comment]
Deploy preview for *giveth-dapp* processing.

Built with commit 5da8561210d6320169da585394ec92f3c2b98652

https://app.netlify.com/sites/giveth-dapp/deploys/5ae37ed111b73b0dadd0ec28

**Comment by vojtechsimetka** [comment]
Deploy preview for *giveth-dapp* processing.

Built with commit 5da8561210d6320169da585394ec92f3c2b98652

https://app.netlify.com/sites/giveth-dapp/deploys/5ae37ed23672df050f61bf1e

File: src/components/GoBackButton.jsx [hunk 1]
```diff
@@ -3,8 +3,32 @@ import PropTypes from 'prop-types';
 
 const GoBackButton = ({ history, styleName }) => (
   <div
-    onClick={history.goBack}
-    onKeyPress={history.goBack}
+    onClick={() => {
+      // ths back button will go a lower nested route inside of the DApp
+      // if the browser history is zero
+      //  this should be the default behavior of the back button
+      if (history.length < 1) {
```
Comment (vojtechsimetka) [inline src/components/GoBackButton.jsx]: Could you extract the logic that is repeated in both `onClick` and `onKeyPress` handlers to a function? Also I don't think this solves the issue of accessing giveth from external site. I think the back button should always go one route lower and if the user wants to go back they use the browser back button.

File: src/lib/middleware.jsx [hunk 2]
```diff
@@ -16,7 +16,7 @@ import { history } from '../lib/helpers';
 export const isLoggedIn = currentUser =>
   new Promise(resolve => {
     if (currentUser && currentUser.address) resolve();
-    else history.goBack();
+    else history.push('/');
```
Comment (vojtechsimetka) [inline src/lib/middleware.jsx]: This is super ugly to have it in the isLoggedIn user. Could this just reject the promise and then the functionality be handled outside? I may even make sense to inform the user that they were accessing protected URL and should login first. What do you think?

File: src/lib/middleware.jsx [hunk 3]
```diff
@@ -35,7 +35,7 @@ export const isLoggedIn = currentUser =>
 export const isAuthenticated = (currentUser, wallet) =>
   new Promise(resolve => {
     if (currentUser && currentUser.address && wallet && wallet.unlocked) resolve();
-    else history.goBack();
+    else history.push('/');
```
Comment (vojtechsimetka) [inline src/lib/middleware.jsx]: Same feedback as in isLoggeIn function.

File: src/components/GoBackButton.jsx (hunk 1 above)
Comment (RnbWd) [reply src/components/GoBackButton.jsx]: yes i agree, I can extract the logic into a utility function, without the conditional statement which isn't necessary

File: src/lib/middleware.jsx (hunk 2 above)
Comment (RnbWd) [reply src/lib/middleware.jsx]: Yes, I agree that makes a lot of sense, but in the meantime this is a better solution

**Comment by RnbWd** [comment]
> @vojtechsimetka I've noticed one more thing, there is a back button when creating/proposing/editing milestone, however that one does not have a campaign in the URL. Could you change the URLs as well?

I noticed that the milestone path in the main application route was structured like that, but it also looked like you could edit and create milestones from a url with a campaign, and that's what I saw when testing, are both paths for milestones identical?

**Comment by vojtechsimetka** [comment]
Yes please do

**Comment by vojtechsimetka** [comment]
@RnbWd are you still working on this? Is there something we can help you with?

**Comment by RnbWd** [comment]
Yes, the changes are pretty simple, but I was wondering what we'd use to notify the user they were accessing a protected URL, a swal popup?

**Comment by vojtechsimetka** [comment]
Hi @RnbWd, is this something you are still actively pursuing?

**Comment by RnbWd** [comment]
I added a popup that either redirects home or to the sign in page if the user is not logged it, and it rejects the promise. I haven't gone through the codebase yet to make sure the promise rejections are handled properly. This PR also changes the functionality of the back button (to go back one lower nest in the url instead of using the browser's default goback behavior)

**Comment by RnbWd** [comment]
the promises are handled okay.. I the popup dialogue doesn't open in the catch statement, so the promise rejection doesn't do much, but its there in case anyone wants to use that function

**Comment by RnbWd** [comment]
made a new pull request https://github.com/Giveth/giveth-dapp/pull/403

**Comment by satyavh** [comment]
Merged this, some feedback:

I reviewed this PR but  the solution only works if the user is actually signed out. If the user is signed in, and you do a browser refresh or go to a restricted page directly, this solution fails.

1. The issue is that when you are signed in, and go to for example `/my-dacs`, the popup shows, but at the same time in the background it loads the DACs.  I'm signed in, so why does the popup show at all? It should not.

The problem is not this PR though (although it should have been tackled). There's a widespread issue within the dApp that re-renders and remounts components. For example, for the MyDACs component, `componentDidMount` is triggered 4 (!) times. That results in this code not working properly:

`isLoggedIn(this.props.currentUser)` 

Because the currentUser is undefined for the first, second and third mount, but the 4th mount the user is defined. However the first mount React.swal is already triggered, and then for the 4th mount the data is loaded in the background.

I've made a fix for that in the `UserProvider` and now the popup works correctly. Yeah! :-)

2. When you are on a restricted route, and press the sign out button, you get to see the pop-up. That's unnecessary information. So I've made a fix to route away from the page.

Apart from these 2 issues, great job! The new implementation of the GoBackButton is helping a lot with other issues we're having. So thanks a lot for tackling this!!

**Pull request closed without merging**